common/beerocks/bcl/source/network/network_utils.cpp

```diff
@@ -363,6 +365,95 @@ static int parseRoutes(struct nlmsghdr *nlHdr, std::shared_ptr<route_info> rtInf
     return (0);
 }
 
+std::shared_ptr<std::unordered_map<std::string, std::string>>
+network_utils::get_arp_table(bool mac_as_key)
+{
+    auto arp_table_ptr = std::make_shared<std::unordered_map<std::string, std::string>>();
+    auto arp_table     = *arp_table_ptr.get();
+
+    // Allocate a new netlink socket
+    auto nl_socket = nl_socket_alloc();
+    if (nl_socket == nullptr) {
+        LOG(ERROR) << "Failed allocating netlink socket!";
+        return nullptr;
+    }
+
+    int err = nl_connect(nl_socket, NETLINK_ROUTE);
+    if (err != 0) {
+        LOG(ERROR) << "Failed connecting the netlink socket: " << nl_geterror(err);
+        nl_socket_free(nl_socket);
+        return nullptr;
+    }
+
+    // ARP table cache
+    struct nl_cache *neighbor_table_cache = nullptr;
+
+    // Allocate the cache and fill it with data
+    err = rtnl_neigh_alloc_cache(nl_socket, &neighbor_table_cache);
+    if (err != 0) {
+        LOG(ERROR) << "Failed probing ARP table: " << nl_geterror(err);
+        nl_socket_free(nl_socket);
+        return nullptr;
+    }
+
+    nl_socket_free(nl_socket);
+
+    auto enteries_count = nl_cache_nitems(neighbor_table_cache);
+
+    constexpr const char *error_message = "neighbor_entry is null";
+    auto neighbor_entry = reinterpret_cast<rtnl_neigh *>(nl_cache_get_first(neighbor_table_cache));
+    if (!neighbor_entry) {
+        LOG(ERROR) << error_message;
+        nl_cache_free(neighbor_table_cache);
+        return nullptr;
+    }
+
+    char buffer_mac[MAC_ADDR_CHAR_SIZE + 1]; // +1 for null terminator
+    char buffer_ip[IPV4_ADDR_CHAR_SIZE + 1]; // +1 for null terminator
+
+    for (int i = 0; i < enteries_count; i++) {
+
+        // Skip NOARP state and non IPv4 records
+        if (rtnl_neigh_get_state(neighbor_entry) == NUD_NOARP ||
+            rtnl_neigh_get_family(neighbor_entry) != AF_INET) {
+            continue;
+        }
+
+        auto nl_addr_mac = rtnl_neigh_get_lladdr(neighbor_entry);
+        if (nl_addr_mac == nullptr) {
+            continue;
+        }
+
+        auto nl_addr_ip = rtnl_neigh_get_dst(neighbor_entry);
+        if (nl_addr_ip == nullptr) {
+            continue;
+        }
+
+        char *mac = nl_addr2str(nl_addr_mac, buffer_mac, sizeof(buffer_mac));
```

Comment:
Isn't it possible to initialize `std::string` with the return value?
```suggestion
        std::string mac = nl_addr2str(nl_addr_mac, buffer_mac, sizeof(buffer_mac));
```

Reply:
What is the benefit of that? I anyway doing construction of the string when inserting it to the map. So doing it will just cause extra string object construction which we don't need.

Reply:
> What is the benefit of that?

It is probably safer as you don't deal with limited buffer size.
It is also removes the need to declare the local buffer (mac and ip) with that +1 size added to the two `#define`s that with `std::string` are not needed at all. The +1 is a source for endless bugs that are simply omitted by using std::string instead of raw pointers.
All in all - uplifting form the bare pointers is safer, cleaner and better.

> doing construction of the string when inserting it to the map. So doing it will just cause extra string object construction which we don't need

Simply use: `arp_table[mac] = std::move(ip)` - no copy at all (and don't use `ip` afterwords).

Reply:
The NL function puts a null terminator after it finishes writing the data into the buffer. It has nothing to do with the function return value.